blockchain/chain.go

```diff
@@ -742,16 +715,65 @@ func (b *BlockChain) getReorganizeNodes(node *blockNode) (*list.List, *list.List
 
 // connectBlock handles connecting the passed node/block to the end of the main
 // (best) chain.
+//
+// This function MUST be called with the chain state lock held (for writes).
 func (b *BlockChain) connectBlock(node *blockNode, block *btcutil.Block) error {
 	// Make sure it's extending the end of the best chain.
 	prevHash := &block.MsgBlock().Header.PrevBlock
-	if b.bestChain != nil && !prevHash.IsEqual(b.bestChain.hash) {
+	if b.bestNode != nil && !prevHash.IsEqual(b.bestNode.hash) {
 		return fmt.Errorf("connectBlock must be called with a block " +
 			"that extends the main chain")
 	}
 
-	// Insert the block into the database which houses the main chain.
-	_, err := b.db.InsertBlock(block)
+	// Generate a new best state snapshot that will be used to update the
+	// database and later memory if all database updates are successful.
+	b.stateLock.RLock()
+	curTotalTxns := b.stateSnapshot.TotalTxns
+	b.stateLock.RUnlock()
+	numTxns := uint64(len(block.MsgBlock().Transactions))
+	blockSize := uint64(block.MsgBlock().SerializeSize())
+	state := newBestState(node, blockSize, numTxns, curTotalTxns+numTxns)
+
+	// Atomically insert info into the database.
+	err := b.db.Update(func(dbTx database.Tx) error {
+		// Update best block state.
+		err := dbPutBestState(dbTx, state, node.workSum)
+		if err != nil {
+			return err
+		}
+
+		// Add the block hash and height to the block index which tracks
+		// the main chain.
+		err = dbPutBlockIndex(dbTx, block.Sha(), node.height)
+		if err != nil {
+			return err
+		}
+
+		// Add transaction index entries for all of the transactions in
+		// the block being connected.
+		err = dbAddTxIndexEntries(dbTx, block)
```

Comment:
Once the pruned utxo set has been implemented this can be made optional to a degree with a `--txidx` flag. And as we discussed on IRC it would be wise to consolidate all indexing operations (`tx\addr`) in a single batched atomic write. 

Leaving this comment here as a reminder.